### ISSUE TITLE
Why rescues are no-blind rescues are enforced and blind returns are not?

### ISSUE THREAD

**Issue by ixti** [rubocop-hq/rubocop#4851]
I'm really curios of reasoning why implicit rescues are considered bad, but at the same time implicit return is considered good?

First of all I believe cop should be renamed to `Lint/RescueStandardErrorClass` and it whould be either forcing to specify StandardError in case of blind rescues, or forcing the opposite - removing StandardError. And by default it should prefer blind style.

Otherwise, it's absolutely not obvious why at the same time we enforce avoiding `return`s.

**Comment by Drenmi** [comment]
> I'm really curios of reasoning why implicit rescues are considered bad [...]

Because it leaves one wondering whether the author actually intended to rescue `StandardError`, or if they just didn't bother to think about what they wanted to rescue. (The latter is more common.)

Reporting an offense forces thinking about that intent at write time, and explicating the error class expresses it at read time. 🙂

(It's notable that the Ruby Style Guide lists both styles as "good". This discussion hadn't been had when the entry was added.)

> [...] it whould be either forcing to specify StandardError in case of blind rescues, or forcing the opposite - removing StandardError. 

RuboCop is heading towards full configurability, so this is a reasonable feature request. I will treat this issue as a request for that configuration option. (There is an existing issue open about explicit `return` keywords.)

> And by default it should prefer blind style.

This discussion is better to be had over at the [Ruby Style Guide](https://github.com/bbatsov/ruby-style-guide). (You'd likely need to construct an argument to counter the first point.)

**Comment by ixti** [comment]
Mmm.... I was in fact just curious. Not having any troubles - just disabling this cop for myself. Just wanted to drag a bit of attention to the fact that if blind rescue:

>  leaves one wondering whether the author actually intended to rescue StandardError, or if they just didn't bother to think about what they wanted to rescue

then one will be left wondering whenever the author actually intended to return last statement or if they didn't bother to think about what they wanted to return in case of implicit return as well.

In addition to that IMO "blind rescue" is a misleading name as in fact it's an implicit rescue from StandardError. Just like implicit return of last expression result.

**Comment by Drenmi** [comment]
> then one will be left wondering whenever the author actually intended to return last statement or if they didn't bother to think about what they wanted to return in case of implicit return as well.

To me the difference is in frequency. I've never seen anyone who didn't think about the actual conditional paths of the method they're implementing. To clarify: I often see people miss out edge cases, but I don't see people accidentally adding edge cases by "unintentionally returning."

I do however quite often see people intentionally returning, but not thinking about what it is they are returning. (Since you can implicitly return `nil` by omitting the return value. There's also a cop for that.) This is the true analogy to the problem with rescuing.

> Thus IMO "blind rescue" is a misleading name as in fact it's an implicit rescue from StandardError. Just like implicit return of last expression result.

Yeah. There hasn't been a consistent naming scheme from the get-go. Thus some cops are named after the "not desirable" style, some after the "desirable" style, and some neutrally.

**Comment by timdiggins** [comment]
@Drenmi just picking up on this. I don't see any argument against "blind rescue" in the style guide at all, just a very valid argument ["Avoid rescuing the Exception class"](https://github.com/bbatsov/ruby-style-guide#no-blind-rescues).

Why does Rubocop need an opinion on this when the ruby style guide doesn't have one?

I note that the title to the section used to be "no-blind-rescues" and indeed, back in 2012 "no blind rescue" was changed from being "still bad" to "good" -- because of misconceptions: https://github.com/bbatsov/ruby-style-guide/commit/1b0fd8f6b602babc992d25b6edd7753cdbb26e27

So I think that rubocop has some thinking in it that has been superseded some 5 years ago in the style guide -- maybe because of the (correctly) preserved permalink... 

Any chance of a reversal on this?

As a rider to this, there are some really good examples of blind exceptions:

It is really common to do:

```ruby
begin 
  do something
rescue => e
  log_something_fairly_specific_about(e)
  raise
end
```

This seems like good code that shouldn't need an explicit exemption. Adding "StandardError" just adds noise (and encourages people to think that a simple `rescue` is somehow bad -- but actually it's what you do inside the rescue that counts).

**Comment by Drenmi** [comment]
> Why does Rubocop need an opinion on this when the ruby style guide doesn't have one?

I think there might be a misconception about what RuboCop is here. It is not the style guide. It is just a tool for enforcing whatever style suits your project. Yes. It was originally built to enforce the style guide, but has since moved towards full configurability.

Thus, RuboCop does not have an opinions. It has defaults. In the cases where the style guide has two entries in "good", the default is kind of arbitrary, but also insignificant, since it is configurable. Once this feature request is fulfilled, there will also be the option to enforce the omission of `StandardError`.

> Any chance of a reversal on this?

I think this question is based on the assumption above, that RuboCop somehow is a style guide. 🙂

It won't be reversed, but the configurability will be extended so you can enforce either style. And there's always the option to disable it entirely in configuration.

**Comment by timdiggins** [comment]
I really appreciate you taking the time to reply, especially as I think many people perhaps bring up the same issue

> I think there might be a misconception about what RuboCop is here. It is not the style guide. It is just a tool for enforcing whatever style suits your project. Yes. It was originally built to enforce the style guide, but has since moved towards full configurability.

> Thus, RuboCop does not have an opinions. It has defaults. 

I can hear what you're saying, but in fact many projects (especially open source projects that don't have time to construct a set of their own preferences) tend to accept RuboCop's defaults as the default style guide. So RuboCop defaults has become a default style guide, despite what it may *want* to be.

I think there are some options (like single quote vs double quote) where RuboCop has to pick one (because consistency is better than variability) and other ones (like this one) where RuboCop should be not enforcing consistency, but instead just looking for the badness (rescue Exception, or rescue with no code within the rescue stanza).

**But** if I read you right...

> Once this feature request is fulfilled, there will also be the option to enforce the omission of StandardError

... then the default will be to **not** enforce omission of Standard Error. In which case we're in total agreement! 😄  
Can you tell me if I read that correctly? If so, I will have a go at encoding that change (as I want it after all).

*But even if not in agreement on this small issue, thank you and your colleagues for the excellent work on RuboCop!*

**Comment by Drenmi** [comment]
> I think there are some options (like single quote vs double quote) where RuboCop has to pick one (because consistency is better than variability) and other ones (like this one) where RuboCop should be not enforcing consistency, but instead just looking for the badness (rescue Exception, or rescue with no code within the rescue stanza).

Neither single- nor double quotes are "bad", so it's hard to see a distinction here. 🤔 

> [...] then the default will be to not enforce omission of Standard Error.

I don't really want to get into the discussion here (but don't mind it in the style guide repo), but so far some specific arguments have been made for explicit error class, and the arguments for implicit one has mostly been name calling, i.e. "noisy", "ugly", etc. But that easily ends in a stalemate if we just call the implicit style "obscure".

It's not that uncommon that people really want their personal preference to be the RuboCop default, but I don't really know why. Perhaps there is a personal fear that *other people* are qualifying error classes out there? 🙂 (This is part tongue-in-cheek, part genuine curiosity.)

To me it is the same, as I belong to the group which configures RuboCop for projects. As long as the "bad" style isn't the default.

> But even if not in agreement on this small issue, thank you and your colleagues for the excellent work on RuboCop!

You are most welcome. 😊 The work (so far) has been nothing but rewarding for me. Mayhaps you will soon join the contributors club with a PR for this? 😉

**Comment by ixti** [comment]
> Neither single- nor double quotes are "bad", so it's hard to see a distinction here.

It's a matter of taste and definition of what to be considered as "bad". In my opinion single quotes are highly bad, as they lead to pretty inconsistent quotes usage (mixing them up with double quotes when interpolation is needed).

> [...] so far some specific arguments have been made for explicit error class

In my opinion argument (at least the one I heard earlier in this thread) is pretty weak. Same can be said about `return`, that omitting latter one leaves one wondering if the method returns something or not.

I would recommend that abstract developer, who is having hard times understanding what is handled by `rescue` without explicit error class RTFM. After all at bare minimum explicit StandardError:

- makes code polluted with useless information that only makes reading of code harder
- encourages to use exceptions as flow control
- encourages to raise/fail with `StandardError` in cases where custom error class is more appropriate
- encourages to raise/fail with `StandardError` instead of raise without explicit error class in case of runtime errors, or even worse, using explicit `RuntimeError` class in that statement

Last two bullets are a bit more complex than they might seem. And if rubocop by default suggests to explicitly rescue `StandardError`, then it should probably propose to be explicit in `raise`/`fail` calls too, and encourage raise with `RuntimeError` instead of being implicit. And probably also show a disclaimer with proposal to think about writing Java instead...

PS I don't think that this discussion belongs to ruby style guide. Otherwise it sounds like you are proposing to change ruby style guide in order to conform with rubocop's defaults. Despite that in rubocop's readme says: *Out of the box it will enforce many of the guidelines outlined in the community Ruby Style Guide.*

**Comment by timdiggins** [comment]
I think it's worth separating this discussion into separate areas: 

1) what is "good" and "bad" -- this is what should be discussed and documented in the [Ruby Style Guide](https://github.com/bbatsov/ruby-style-guide) -- it seems relatively slow (good!) and consensual (very good!) in adopting changes there.

2) what rubocop's OOB defaults are -- this discussion should go on evidently in rubocop issues etc (and not Ruby Style Guide). Perhaps there is a page that determines the principles of the relationship between RuboCop and Ruby Style Guide (I could not find one in the wiki --  maybe if we get consensus here I can write a first draft in the wiki).

Regarding 2):
Firstly 
> "Out of the box it will enforce many of the guidelines outlined in the community Ruby Style Guide"

This means (to me) that it will attempt to follow the following rules:

1. **Mark bad**: mark areas that are highlighted as "bad" in the style guide, but sometimes won't be able to (there are limits to what static analysis can do).
2. **Enforce consistency** between two or more alternative "good" styles, where this is deemed important (the default will be arbitrary and configurable ). As an example  Ruby style guid says that string literal consistency is important (https://github.com/bbatsov/ruby-style-guide#consistent-string-literals), but which of double or single quote Rubocop has had to choose one for default in order to enforce the consistency).
3. **Allow for local rules and extensions**: allow code repostories to create and enforce local standards through configuration and extension of Rubocop

The main question that I'm proposing is whether Rubocop out of the box needs to be an aggressive enforcer of things in the third camp, or whether instead it should enforce the first two above, and leave the others turned off by default

**Comment by andy-smith-msm** [comment]
Robocop represents a great tool for us here and one that we really appreciate.

It offered what we believed to be an exact interpretation of the Ruby Style Guide and has helped greatly to suggest standardisation in a non partisan fashion. The second that one starts to implement general override files for personal preference negates that and also means that one loses any real changes in opinion that we believed came from the RSG.

I don't think that I'm unique in finding the suggestion to change a default that used to be perfectly good and hasn't changed because of a change in RSG opinion to be odd and irksome now that previously good code is failing linting checks.

Our approach is to never use override files. Overrides here need to be justified in comments so that they can be considered every time the code is reviewed and one member of the team may have a better take on how to meet the RSG so the comment approach is better and doesn't weaken that message that somebody's opinion in our team might outweigh through personal preference a well thought out argument from the RSG.

**Comment by Drenmi** [comment]
@andy-smith-msm: Would it be helpful to you if there was a RuboCop configuration that's kept on par with the style guide? Thinking out loud here. Perhaps even as a flag in RuboCop itself? E.g.:

```
rubocop --ruby-style-guide
```

**Comment by andy-smith-msm** [comment]
@Drenmi I don't think there's anything unusual about the way we use Rubocop and think that suggestion would be very useful to us and many others.

**Comment by timdiggins** [comment]
@Drenmi @andy-smith-msm I love the idea of being able to reduce the rubocop to a more minimal ruby-style-guide norm. I think having a command line flag (which you can also set within .rubycop.yml) would be very useful. (Then a discussion might be on a new project: -- shall we go with Ruby Style Guide, or shall we go with the Full Rubocop). 

Note however even with Ruby Style Guide approach there are some arbitrary choices that rubocop has had to make for its default (like which string format to choose when mandating consistency) -- this is a decent compromise though.

**Comment by andy-smith-msm** [comment]
@timdiggins I fully agree, there are some areas that Rubocop adds and where we further coach folks to use the language and libraries more expressively with the aim of communicating more intent in the way that the code reads. 
Maybe the RSG doesn't mandate all of that and Rubocop can't necessarily advise on, say, which alias method is more communicative in a particular context but some of the good things such as quote consistency really help to produce a common idiomatic style and that's really useful and appreciated.

**Comment by bbatsov** [comment]
Great discussion! 

Generally a lot of the inconsistencies between the guide and RuboCop came from the fact that often some cops were added without being reflected in the guide and that in the early days of RuboCop it was not intended to be configurable at all. There should definitely be more alignment between the two projects and all cops should be configurable when this makes sense. We're working on this, but it will take us a while go get there.

There hasn't been much activity on this ticket and our Core Team is spread too thin on so many tasks, so I'll just close it for the sake of having a cleaner lists of tasks to focus on. We'd gladly review a PR, but it's unlikely that'd we're going to tackle this ourselves in the foreseeable future.

**Comment by ixti** [comment]
Current state of StandardError style cop is perfectly fine IMO. It allows to enforce one style or the other. Thank you!
